server/src/main/java/org/apache/iotdb/db/conf/adapter/ActiveTimeSeriesCounter.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.conf.adapter;
+
+import com.clearspring.analytics.stream.cardinality.HyperLogLog;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.iotdb.db.conf.IoTDBConstant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ActiveTimeSeriesCounter implements IActiveTimeSeriesCounter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ActiveTimeSeriesCounter.class);
+  /**
+   * Map[StorageGroup, HyperLogLogCounter]
+   */
+  private static Map<String, HyperLogLog> storageGroupHllMap = new ConcurrentHashMap<>();
+
+  /**
+   * Map[StorageGroup, ActiveTimeSeriesRatio]
+   */
+  private static Map<String, Double> activeRatioMap = new ConcurrentHashMap<>();
+
+  /**
+   * Map[StorageGroup, ActiveTimeSeriesNumber]
+   */
+  private static Map<String, Long> activeTimeSeriesNumMap = new ConcurrentHashMap<>();
+
+  /**
+   * LOG2M decide the precision of the HyperLogLog algorithm
+   */
+  static final int LOG2M = 13;
+
+  private ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+  @Override
+  public void init(String storageGroup) {
+    storageGroupHllMap.put(storageGroup, new HyperLogLog(LOG2M));
+    activeRatioMap.put(storageGroup, 0D);
+    activeTimeSeriesNumMap.put(storageGroup, 0L);
+  }
+
+  @Override
+  public void offer(String storageGroup, String device, String measurement) {
+    String path = device + IoTDBConstant.PATH_SEPARATOR + measurement;
+    try {
+      storageGroupHllMap.get(storageGroup).offer(path);
+    } catch (Exception e) {
+      storageGroupHllMap.putIfAbsent(storageGroup, new HyperLogLog(LOG2M));
+      storageGroupHllMap.get(storageGroup).offer(path);
+      LOGGER.error("Register active time series root.{}.{} failed", storageGroup, path, e);
+    }
+  }
+
+  @Override
+  public void updateActiveRatio(String storageGroup) {
+    lock.writeLock().lock();
+    try {
+      long activeTimeSeriesNum = storageGroupHllMap.get(storageGroup).cardinality();
+      if (activeTimeSeriesNum != activeTimeSeriesNumMap.get(storageGroup)) {
```

Comment:
It seems that if the latest `storageGroupHllMap.get(storageGroup).cardinality()` just happens to equal to `activeTimeSeriesNumMap.get(storageGroup)`, the `activeRatio` wouldn't be updated. Is that a bug?

Besides, the logic of the judgment condition - `activeTimeSeriesNum != activeTimeSeriesNumMap.get(storageGroup)` - looks a little bit wierd to me. If you want to judge whether the activeRatio of a storageGroup needs updating, you need to take other storageGroups into consideration. But the judgment condition here looks more like a decoration since the HLL counter is initialized every time in this function and thus the new counted result has a high probability of not equaling to `activeTimeSeriesNumMap.get(storageGroup)` (but still may equal hence the bug mentioned above, right?).

Reply:
If the activeTimeSeriesNum is not changed then the activeTimeSeriesNumMap is no need to be modified.